multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowBranchProjectFactoryTest.java

```diff
@@ -0,0 +1,100 @@
+import org.junit.Rule;
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+
+public class WorkflowBranchProjectFactoryTest {
+
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+    @Rule public GitSampleRepoRule sampleRepo = new GitSampleRepoRule();
+
+    @Issue("JENKINS-30744")
+    @Test public void slashyBranches() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                sampleRepo.init();
+                sampleRepo.git("checkout", "-b", "dev/main");
+                ScriptApproval.get().approveSignature("method java.lang.String replaceFirst java.lang.String java.lang.String"); // TODO add to generic-whitelist
+                sampleRepo.write("Jenkinsfile",
+                    "echo \"branch=${env.BRANCH_NAME}\"\n" +
+                    "node {\n" +
+                    "  checkout scm\n" +
+                    "  echo \"workspace=${pwd().replaceFirst('.+dev', 'dev')}\"\n" +
+                    "}");
+                sampleRepo.git("add", "Jenkinsfile");
+                sampleRepo.git("commit", "--all", "--message=flow");
+                WorkflowMultiBranchProject mp = story.j.jenkins.createProject(WorkflowMultiBranchProject.class, "p");
+                mp.getSourcesList().add(new BranchSource(new GitSCMSource(null, sampleRepo.toString(), "", "*", "", false), new DefaultBranchPropertyStrategy(new BranchProperty[0])));
+                WorkflowJob p = scheduleAndFindBranchProject(mp, "dev%2Fmain");
+                assertEquals(1, mp.getItems().size());
+                story.j.waitUntilNoActivity();
+                WorkflowRun b1 = p.getLastBuild();
+                assertEquals(1, b1.getNumber());
+                story.j.assertLogContains("branch=dev/main", b1);
+                story.j.assertLogContains("workspace=dev%2Fmain", b1);
+                verifyProject(p);
+            }
+        });
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("p/dev%2Fmain", WorkflowJob.class);
+                assertNotNull(p);
+                WorkflowRun b2 = p.scheduleBuild2(0).get();
+                assertEquals(2, b2.getNumber());
+                story.j.assertLogContains("branch=dev/main", b2);
+                story.j.assertLogContains("workspace=dev%2Fmain", b2);
+                verifyProject(p);
+            }
+        });
+    }
+    private void verifyProject(WorkflowJob p) throws Exception {
+        assertEquals("dev%2Fmain", p.getName());
+        assertEquals("dev/main", p.getDisplayName());
+        assertEquals("p/dev%2Fmain", p.getFullName());
+        assertEquals("p » dev/main", p.getFullDisplayName());
```

Comment:
As mentioned in the PR description I ran this test on Windows (XP, if it matters) and it passed, so there is not a filesystem problem. `/` is the only special character I have heard of from the field (used in Git Flow IIUC). `\` is handled by `Util.rawEncode`. If there are other characters which are _not_ handled by `rawEncode` but which pose a problem on some filesystems, that could be a follow-up fix localized to `Branch.getEncodedName` plus `MultiBranchProject.rawDecode`.